ocaml/testsuite/tests/typing-modes/module.ml

```diff
@@ -0,0 +1,88 @@
+(* TEST
+   expect;
+*)
+
+(* This file tests that modules are sound wrt modes. *)
+
+module type S = sig val x : string end
+
+module type Empty = sig end
+
+module M = struct
+    let x = "string"
+end
+[%%expect{|
+module type S = sig val x : string end
+module type Empty = sig end
+module M : sig val x : string end
+|}]
+
+(* Closing over modules affects closure's modes *)
+let (foo @ portable) () =
+    let _ = (module M : S) in
+    ()
+(* CR zqian: This should fail *)
+[%%expect{|
+val foo : unit -> unit = <fun>
+|}]
+
+(* File-level modules are looked up differently and need to be tested
+separately. *)
+let (foo @ portable) () =
+    let _ = (module List : Empty) in
+    ()
+(* CR zqian: this should fail *)
+[%%expect{|
+val foo : unit -> unit = <fun>
+|}]
+
+(* Values in modules are defined as legacy *)
+module M = struct
+    let x = local_ "hello"
+end
+[%%expect{|
+Line 2, characters 8-9:
+2 |     let x = local_ "hello"
+            ^
+Error: This value escapes its region.
+|}]
+
+(* Values from modules are available as legacy *)
+let (foo @ portable) () = M.x
```

Comment:
This test will pass once `string` mode-crosses portability. Maybe make the module element a function to protect against that and keep the test robust?